### PR TITLE
Add include-what-you-use

### PR DESCRIPTION
**Pull request by xmnlab**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [ ] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by xmnlab** [comment]
@conda-forge/help-c-cpp this PR is ready for review. thanks!!

File: recipes/iwyu/build.sh [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env bash
+
+CC=clang
+CXX=clang++
```
Comment (xhochy) [inline recipes/iwyu/build.sh:4]: Shouldn't this be the default compilers?  Why is there a need to override them?

File: recipes/iwyu/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    - {{ compiler('cxx') }}
+    - cmake
+    - python
+    - make
+  host:
+    - {{ compiler('cxx') }}
+    - clangdev {{ llvm_version }}
+    - llvmdev {{ llvm_version }}
+    - llvm {{ llvm_version }}
+    - libclang {{ llvm_version }}
```
Comment (xhochy) [inline recipes/iwyu/meta.yaml:29]: The build log suggests that this is not needed.
```suggestion
```

File: recipes/iwyu/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+requirements:
+  build:
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+    - cmake
+    - python
+    - make
+  host:
+    - {{ compiler('cxx') }}
```
Comment (xhochy) [inline recipes/iwyu/meta.yaml:25]: What is the reason you also list this in `host`?

File: recipes/iwyu/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,53 @@
+{% set name = "include-what-you-use" %}
+{% set version = "0.17" %}
+{% set llvm_version = "13.0.1" %}
```
Comment (xhochy) [inline recipes/iwyu/meta.yaml:3]: From the website it looks like the major version should be sufficient. Any reason to pin stronger here?

File: recipes/iwyu/build.sh (hunk 1 above)
Comment (xmnlab) [reply recipes/iwyu/build.sh:4]: no reason at all, normally I prefer to use clang over gcc, but actually there is no specific reason for that.

File: recipes/iwyu/meta.yaml (hunk 2 above)
Comment (xmnlab) [reply recipes/iwyu/meta.yaml:29]: thanks I will remove that!

File: recipes/iwyu/meta.yaml (hunk 3 above)
Comment (xmnlab) [reply recipes/iwyu/meta.yaml:25]: it was not working locally without that ... maybe using gcc would work without that.

File: recipes/iwyu/meta.yaml (hunk 4 above)
Comment (xmnlab) [reply recipes/iwyu/meta.yaml:3]: not sure if I am following you here, I followed this table here: https://github.com/include-what-you-use/include-what-you-use#how-to-build
could you explain a bit more about that pls?

**Comment by xmnlab** [comment]
@xhochy I applied your suggestions and it is working :) thank you so much
just waiting for clarification about https://github.com/conda-forge/staged-recipes/pull/18539#discussion_r846726607
thank you so much for your review.

File: recipes/iwyu/meta.yaml (hunk 4 above)
Comment (xmnlab) [reply recipes/iwyu/meta.yaml:3]: @xhochy just a friendly reminder about this :)

**Comment by xmnlab** [comment]
@conda-forge/help-c-cpp just a friendly reminder about this PR :) thanks

File: recipes/iwyu/build.sh [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  -G "Unix Makefiles" \
+  -DCMAKE_INSTALL_PREFIX=${PREFIX} \
+  -DCMAKE_PREFIX_PATH=${PREFIX} \
+  -DCMAKE_BUILD_TYPE=release \
+  -DCMAKE_C_COMPILER=${CC} \
+  -DCMAKE_CXX_COMPILER=${CXX} \
+  ..
+
+cmake --build .
+cmake --install . -v
```
Comment (ocefpaf) [inline recipes/iwyu/build.sh:15]: Is there a ctest or make check we can run here?

Comment (xmnlab) [reply recipes/iwyu/build.sh:16]: good catch @ocefpaf ! it seems that ctest is available and it seems that some tests are failing. I am investigating that now. thank you so much! really appreciate that :)

**Review by xhochy — APPROVED** [review]
My comments have been addressed 👍

File: recipes/iwyu/build.sh (hunk 5 above)
Comment (xmnlab) [reply recipes/iwyu/build.sh:16]: it is raising weird issue about the standard library access https://github.com/include-what-you-use/include-what-you-use/issues/1033#issuecomment-1107884014

any idea about that?

Comment (ocefpaf) [reply recipes/iwyu/build.sh:16]: I'm not sure I follow the upstream comment, do you need a c++ compiler in the test environment?

Comment (xmnlab) [reply recipes/iwyu/build.sh:16]: thanks @ocefpaf for your attention here :)

this is part of the log from ctest:

```
14/143 Testing: cxx.test_badinc
14/143 Test: cxx.test_badinc
Command: "/mnt/sda1/storage/miniconda3/envs/conda-build/conda-bld/include-what-you-use_1650805645046/_build_env/bin/python" "run_iwyu_tests.py" "--run-test-file=tests/cxx/badinc.cc" "--" "/mnt/sda1/storage/miniconda3/envs/conda-build/conda-bld/include-what-you-use_1650805645046/work/build/bin/include-what-you-use"
Directory: /mnt/sda1/storage/miniconda3/envs/conda-build/conda-bld/include-what-you-use_1650805645046/work
"cxx.test_badinc" start time: Apr 24 09:13 -04
Output:
----------------------------------------------------------
INFO:root:Testing iwyu on tests/cxx/badinc.cc
>>> Running /mnt/sda1/storage/miniconda3/envs/conda-build/conda-bld/include-what-you-use_1650805645046/work/build/bin/include-what-you-use -Xiwyu --verbose=3 -Xiwyu --mapping_file=tests/cxx/badinc.imp -I .  tests/cxx/badinc.cc
tests/cxx/badinc.cc:60:10: fatal error: 'typeinfo' file not found
#include <typeinfo>   // for typeid
         ^~~~~~~~~~
```

I am also new to iwyu ... so maybe it is also a lack of understanding about how to test it XD
but it seems that it depends on the compiler for running as well

Comment (xmnlab) [reply recipes/iwyu/build.sh:16]: mmm maybe I would need to patch the tests: https://github.com/include-what-you-use/include-what-you-use/blob/master/iwyu_tool_test.py#L103
or maybe I could just skip these "problematic" tests ..
what is the CF common approach for that?

Comment (ocefpaf) [reply recipes/iwyu/build.sh:16]: If you can make a case that those test do not matter you can skip them for now, get this merged, and open an issue to track them in the feedstock.

Comment (carterbox) [reply recipes/iwyu/build.sh:16]: You can use `ctest --output-on-failure` to show why the tests fail, and you can also use `-R` to filter the failing tests with regular expressions.

https://cmake.org/cmake/help/latest/manual/ctest.1.html

Comment (xmnlab) [reply recipes/iwyu/build.sh:16]: @ocefpaf I am checking with the iwyu community about this problem ... I am not sure yet if it is a big deal or not

Comment (xmnlab) [reply recipes/iwyu/build.sh:16]: I isolated one test and ran it with more flags, and it worked fine:

```
export iwyu=/mnt/sda1/storage/miniconda3/envs/conda-build/conda-bld/include-what-you-use_1651536695168/work/build/bin/include-what-you-use

$iwyu -Xiwyu --verbose=3 \
  -I${PREFIX}/include \
  -I${BUILD_PREFIX}/x86_64-conda-linux-gnu/include/c++/11.2.0/ \
  -I${BUILD_PREFIX}/x86_64-conda-linux-gnu/include/c++/11.2.0/x86_64-conda-linux-gnu/ \
  -I . tests/cxx/uses_printf.cc
  ```

so I guess that it is safe to skip the test that are failing for now, and try to fix that in the feedstock

**Comment by xmnlab** [comment]
@ocefpaf @carterbox this PR is ready for another review :) thank you so much!

**Comment by xmnlab** [comment]
thanks for the review @ocefpaf @carterbox @xhochy !!! <3

**Pull request merged**